src/commands_upstream.c

```diff
@@ -0,0 +1,608 @@
+        void const *data = NULL;
+
+        while (l_genl_attr_next(&attr, &type, &len, &data)) {
+                if (type != MPTCP_PM_ATTR_ADDR) {
+                        // Sanity check.  This condition should never occur.
+                        l_error("get_addr: unexpected data");
+                        continue;
+                }
+
+                size_t const offset = addrs_len++;
+
+                addrs = l_realloc(addrs, sizeof(*addrs) * addrs_len);
+
+                if (!get_addr_callback_recurse(&attr, addrs + offset)) {
+                        l_free(addrs);
+                        addrs = NULL;
+                        addrs_len = 0;
+                        break;
+                }
+        }
+
+get_addr_out:
+
+        // Pass the results to the user.
+        if (cb->get_addr)
+                cb->get_addr(addrs, cb->data);
+        else
+                cb->dump_addrs(addrs, addrs_len, cb->data);
+
+        l_free(addrs);
+}
+
+static bool append_addr_attr(struct l_genl_msg *msg,
+                             struct sockaddr const *addr)
+{
+        assert(mptcpd_is_inet_family(addr));
+
+        uint16_t type = 0;
+        uint16_t len = 0;
+        void const *data = NULL;
+
+        if (addr->sa_family == AF_INET) {
+                type = MPTCP_PM_ADDR_ATTR_ADDR4;
+
+                struct sockaddr_in const *const addr4 =
+                        (struct sockaddr_in *) addr;
+
+                data = &addr4->sin_addr;
+                len  = sizeof(addr4->sin_addr);
+        } else {
+                type = MPTCP_PM_ADDR_ATTR_ADDR6;
+
+                struct sockaddr_in6 const *const addr6 =
+                        (struct sockaddr_in6 *) addr;
+
+                data = &addr6->sin6_addr;
+                len  = sizeof(addr6->sin6_addr);
+        }
+
+        return l_genl_msg_append_attr(msg, type, len, data);
+}
+
+// --------------------------------------------------------------
+
+static void get_limits_callback(struct l_genl_msg *msg, void *user_data)
+{
+        struct get_limits_user_callback const *const cb = user_data;
+        struct mptcpd_limit *limits = NULL;
+        size_t limits_len = 0;
+
+        if (!mptcpd_check_genl_error(msg, "get_limits"))
+                goto get_limits_out;
+
+        struct l_genl_attr attr;
+        if (!l_genl_attr_init(&attr, msg)) {
+                l_error("get_limits: "
+                        "Unable to initialize genl attribute");
+
+                goto get_limits_out;
+        }
+
+        /*
+          Payload (nested):
+              Network address
+              Address ID
+              Flags
+              Network
+        */
+
+        uint16_t type;
+        uint16_t len;
+        void const *data = NULL;
+
+        while (l_genl_attr_next(&attr, &type, &len, &data)) {
+                size_t const offset = limits_len++;
+
+                limits = l_realloc(limits,
+                                   sizeof(*limits) * limits_len);
```

Comment:
We don't end up doing a lot of iterations here, right? Shouldn't be too much overhead from reallocating, and I don't think this will run that often.

Reply:
Right.  There are currently only two types of limits defined by the upstream kernel's PM netlink API.  At most this loop will currently only have a maximum of two iterations.  I suppose we could just pre-allocate an array with two elements rather than reallocating like this, and just set the length accordingly, but I don't like the idea of assuming the length will always be a maximum of two.  Alternatively we could iterate through the netlink attributes twice, the first time to determine the number of MPTCP limits retrieved from the kernel and allocate the array afterward, and the second time to populate the array.   Thoughts?